### PR TITLE
Adapt to the new version of playerfactions

### PR DESCRIPTION
**Pull request by Kalio-42**
* Enable to open an area to many factions
* Update to adapt to the new version of playerfactions mod, with its new multi-faction mode

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [LuaCheck](https://github.com/minetest-mods/areas/runs/1008685604) warnings.

File: api.lua [hunk 1]
```diff
@@ -92,9 +92,21 @@ function areas:canInteract(pos, name)
 		if area.owner == name or area.open then
 			return true
 		elseif areas.factions_available and area.faction_open then
-			local faction_name = factions.get_player_faction(area.owner)
-			if faction_name ~= nil and faction_name == factions.get_player_faction(name) then
-				return true
+			local faction_name = nil
+			if factions.version == nil or factions.version < 2 then
+				faction_name = factions.get_player_faction(area.owner)
```
Comment (louisroyer) [inline api.lua]:
```suggestion
			if factions.version == nil or factions.version < 2 then
				local faction_name = factions.get_player_faction(area.owner)
```

File: api.lua [hunk 2]
```diff
@@ -92,9 +92,21 @@ function areas:canInteract(pos, name)
...
+				if faction_name ~= nil and faction_name == factions.get_player_faction(name) then
+					return true
+				end
+			else
+				factions_names = area.faction_open
```
Comment (louisroyer) [inline api.lua]:
```suggestion
				local factions_names = area.faction_open
```

**Review by louisroyer — CHANGES_REQUESTED** [review]

File: api.lua (hunk 2 above)
Comment (Kalio-42) [reply api.lua]: Thanks

**Comment by Kalio-42** [comment]
> Please fix [LuaCheck](https://github.com/minetest-mods/areas/runs/1008685604) warnings.

Done

**Review by Panquesito7 — APPROVED** [review]
Looks good to me. 👍
@SmallJoker please review. Thanks.

File: api.lua [hunk 3]
```diff
@@ -92,9 +92,20 @@ function areas:canInteract(pos, name)
...
-				return true
+			if factions.version == nil or factions.version < 2 then
+				local faction_name = factions.get_player_faction(area.owner)
+				if faction_name ~= nil and faction_name == factions.get_player_faction(name) then
+					return true
+				end
+			else
+				local factions_names = area.faction_open
+				if factions_names ~= nil then
+					for _, fname in ipairs(factions_names) do
```
Comment (SmallJoker) [inline api.lua]: Can be shortened to
`for _, fname in ipairs(area.faction_open or {}) do`

File: chatcommands.lua [hunk 4]
```diff
@@ -286,10 +286,15 @@ minetest.register_chatcommand("area_open", {
...
 	minetest.register_chatcommand("area_faction_open", {
-		params = S("<ID>"),
+		params = S("<ID> [faction_name]"),
 		description = S("Toggle an area open/closed for members in your faction."),
 		func = function(name, param)
-			local id = tonumber(param)
+			local params = {}
+			for p in string.gmatch(param, "[^%s]+") do
+				table.insert(params, p)
+			end
```
Comment (SmallJoker) [inline chatcommands.lua]: `params = param:split(" ")`

Not tested.

File: chatcommands.lua [hunk 5]
```diff
@@ -298,12 +303,41 @@ if areas.factions_available then
...
-			return true, open and S("Area opened for faction members.")
-				or S("Area closed for faction members.")
+			if factions.version == nil or factions.version < 2 or factions.mode_unique_faction then
+				local open = not areas.areas[id].faction_open and {factions.get_player_faction(name)}
+				-- Save false as nil to avoid inflating the DB.
+				areas.areas[id].faction_open = open or nil
+				areas:save()
+				return true, open and S("Area opened for faction members.")
+					or S("Area closed for faction members.")
+			else
```
Comment (SmallJoker) [inline chatcommands.lua:314]: Superfluous else. Replace by `end` to reduce the indent levels below.

File: chatcommands.lua [hunk 6]
```diff
@@ -298,12 +303,41 @@ if areas.factions_available then
...
+				local faction_name = params[2]
+				if not factions.get_owner(faction_name) then
+					return false, S("Faction doesn't exists")
+				end
+				local fnames = areas.areas[id].faction_open
+				if fnames == nil then
+					fnames = {}
+				end
```
Comment (SmallJoker) [inline chatcommands.lua]:
```suggestion
				local fnames = areas.areas[id].faction_open or {}
```

File: chatcommands.lua [hunk 7]
```diff
@@ -298,12 +303,41 @@ if areas.factions_available then
...
+				local removed = false
+				for i, fac_name in ipairs(fnames) do
+					if faction_name == fac_name then
+						table.remove(fnames,i)
+						removed = true
+					end
+				end
+				if not removed then
+					table.insert(fnames,faction_name)
```
Comment (SmallJoker) [inline chatcommands.lua]:
```suggestion
				local pos = table.indexof(fnames, faction_name)
				if pos < 0 then
					-- Add new faction to the list
					table.insert(fnames, faction_name)
				else
					table.remove(fnames, pos)
				end
```

File: chatcommands.lua [hunk 8]
```diff
@@ -298,12 +303,41 @@ if areas.factions_available then
...
+			if factions.version == nil or factions.version < 2 or factions.mode_unique_faction then
+				local open = not areas.areas[id].faction_open and {factions.get_player_faction(name)}
+				-- Save false as nil to avoid inflating the DB.
+				areas.areas[id].faction_open = open or nil
+				areas:save()
+				return true, open and S("Area opened for faction members.")
+					or S("Area closed for faction members.")
+			else
+				local faction_name = params[2]
+				if not factions.get_owner(faction_name) then
```
Comment (SmallJoker) [inline chatcommands.lua]: What if the faction was removed in the meantime? You could not remove the entry from the area's access list.

File: chatcommands.lua [hunk 9]
```diff
@@ -298,12 +303,41 @@ if areas.factions_available then
 				return false, S("Area @1 does not exist"
 						.." or is not owned by you.", id)
 			end
-			local open = not areas.areas[id].faction_open
-			-- Save false as nil to avoid inflating the DB.
-			areas.areas[id].faction_open = open or nil
-			areas:save()
-			return true, open and S("Area opened for faction members.")
-				or S("Area closed for faction members.")
+			if factions.version == nil or factions.version < 2 or factions.mode_unique_faction then
```
Comment (SmallJoker) [inline chatcommands.lua]:
```suggestion
			if (factions.version or 0) < 2 or factions.mode_unique_faction then
```

File: chatcommands.lua (hunk 8 above)
Comment (Kalio-42) [reply chatcommands.lua]: In hud.lua I add a verification to remove removed factions, to display only existing faction (and it saves areas if a faction is removed) but i can add the same check here if you want

Comment (SmallJoker) [reply chatcommands.lua]: No, I think it's fine to have it in one place only. Not sure whether HUDs are the right location but acceptable if it works like that.

**Comment by SmallJoker** [comment]
This is a master branch, thus I cannot push to it. Would you please be so nice and apply following patch and check whether it looks alright and still works?
<details>
<summary>Long patch here</summary>

```patch
From 24fa5c60f22ddf41082997a837a63bf038903031 Mon Sep 17 00:00:00 2001
From: SmallJoker <mk939@ymail.com>
Date: Tue, 25 Aug 2020 10:06:50 +0200
Subject: [PATCH] playerfactions support: Consistent formatting, document

---
 .luacheckrc      |  2 +-
 README.md        | 16 +++++++++++++---
 api.lua          |  2 +-
 chatcommands.lua |  8 ++++++--
 hud.lua          | 20 +++++++++++---------
 5 files changed, 32 insertions(+), 16 deletions(-)

diff --git a/.luacheckrc b/.luacheckrc
index c47f79b..eaf3a6e 100644
--- a/.luacheckrc
+++ b/.luacheckrc
@@ -10,7 +10,7 @@ read_globals = {
 	"AreaStore",
 	"default",
 	"factions",
-	table = { fields = { "copy", "getn" } }
+	table = { fields = { "copy", "getn", "indexof" } }
 }
 
 globals = {
diff --git a/README.md b/README.md
index bdfe4b2..1c841c4 100644
--- a/README.md
+++ b/README.md
@@ -4,7 +4,12 @@ Areas mod for Minetest
 Dependencies
 ------------
 
-Minetest 5.0.0+ is recommended, but 0.4.16+ should work as well.
+Minetest 5.0.0+
+
+Optional support for following mods:
+
+  * [playerfactions](https://forum.minetest.net/viewtopic.php?t=23281) by aldenp
+  * [factions](https://forum.minetest.net/viewtopic.php?t=20949) by Coder12
 
 
 Configuration
@@ -107,12 +112,17 @@ Commands
 
   * `/area_open <ID>` -- Toggle open/closed the specified area for everyone.
 
-  * `/area_faction_open <ID> [faction]` -- Toggle open/closed the specified area for members of the faction. Factions are created and managed by playerfactions mod. 
+  * `/area_faction_open <ID> [faction]` -- Toggle open/closed the specified
+	area for members of the faction. Factions are created and managed by
+	playerfactions mod. 
 
 License
 -------
 
-Copyright (C) 2013 ShadowNinja
+Copyright (C) 2013-2017 ShadowNinja
+
+Copyright (C) 2015-2020 various contributors
+
 
 Licensed under the GNU LGPL version 2.1 or later.
 See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
diff --git a/api.lua b/api.lua
index 0a9e146..ffd8be1 100644
--- a/api.lua
+++ b/api.lua
@@ -92,7 +92,7 @@ function areas:canInteract(pos, name)
 		if area.owner == name or area.open then
 			return true
 		elseif areas.factions_available and area.faction_open then
-			if factions.version == nil or factions.version < 2 then
+			if (factions.version or 0) < 2 then
 				local faction_name = factions.get_player_faction(area.owner)
 				if faction_name ~= nil and faction_name == factions.get_player_faction(name) then
 					return true
diff --git a/chatcommands.lua b/chatcommands.lua
index 8def94f..64d0f93 100644
--- a/chatcommands.lua
+++ b/chatcommands.lua
@@ -302,6 +302,7 @@ if areas.factions_available then
 						.." or is not owned by you.", id)
 			end
 			if (factions.version or 0) < 2 or factions.mode_unique_faction then
+				-- Single faction mode
 				local open = not areas.areas[id].faction_open and {factions.get_player_faction(name)}
 				-- Save false as nil to avoid inflating the DB.
 				areas.areas[id].faction_open = open or nil
@@ -309,6 +310,8 @@ if areas.factions_available then
 				return true, open and S("Area opened for faction members.")
 					or S("Area closed for faction members.")
 			end
+
+			-- Multiple factions support
 			local faction_name = params[2]
 			if not factions.get_owner(faction_name) then
 				return false, S("Faction doesn't exists")
@@ -321,13 +324,14 @@ if areas.factions_available then
 			else
 				table.remove(fnames, pos)
 			end
+
 			if #fnames == 0 then
+				-- Save {} as nil to avoid inflating the DB.
 				fnames = nil
 			end
-			-- Save {} as nil to avoid inflating the DB.
 			areas.areas[id].faction_open = fnames
 			areas:save()
-			return true, fnames and S("Area is open for members of: @1",table.concat(fnames,", "))
+			return true, fnames and S("Area is open for members of: @1", table.concat(fnames, ", "))
 				or S("Area closed for faction members.")
 		end
 	})
diff --git a/hud.lua b/hud.lua
index 39b8837..0bfd1ca 100644
--- a/hud.lua
+++ b/hud.lua
@@ -20,28 +20,30 @@ minetest.register_globalstep(function(dtime)
 		local areaStrings = {}
 
 		for id, area in pairs(areas:getAreasAtPos(pos)) do
-			local faction_info = area.faction_open and areas.factions_available
-			if faction_info then
-				if factions.version == nil or factions.version < 2 then
+			local faction_info
+			if area.faction_open and areas.factions_available then
+				-- Gather and clean up disbanded factions
+				if (factions.version or 0) < 2 then
 					faction_info = factions.get_player_faction(area.owner)
 				else
 					-- Verify that every displayed faction still exists
-					local faction_open_changed = false
+					local changed = false
 					for i, fac_name in ipairs(area.faction_open) do
 						if not factions.get_owner(fac_name) then
 							table.remove(area.faction_open, i)
-							faction_open_changed = true
+							changed = true
 						end
 					end
 					if #area.faction_open == 0 then
+						-- Prevent DB clutter, remove value
 						area.faction_open = nil
-						faction_info = nil
-						faction_open_changed = true
 					else
 						faction_info = table.concat(area.faction_open, ", ")
 					end
-					-- Save areas if a faction was disband
-					if faction_open_changed then areas:save()	end
+
+					if changed then
+						areas:save()
+					end
 				end
 			end
 
-- 
2.25.1


```
</details>

File: chatcommands.lua (hunk 8 above)
Comment (Kalio-42) [reply chatcommands.lua]: It works well ;)
Putting it here decrease the size of area.faction_open (so it decrease the number of checks) and display only existing faction.
And it saves areas only if there is some changes.

**Comment by Kalio-42** [comment]
Done, @SmallJoker

**Comment by SmallJoker** [comment]
So you dropped the single-faction mode? Does it still work with "factions" and "playerfactions" mentioned in the README.md?

**Comment by Kalio-42** [comment]
> So you dropped the single-faction mode? Does it still work with "factions" and "playerfactions" mentioned in the README.md?

area_faction_open use only playerfactions mod (I just checked the mod factions you tell me to put in README.md but there was no adaptation to the mod) 
Yes the purpose of this pull request was to adapt to playerfactions so it still work with the old and the new version (that I made).

I made this change because I think it's better for consistency that the version of playerfactions don't modify the utilisation of the areas mod, (I mean that before with the old version of playerfactions we could only open an area to one faction but if we update to the new version we would be abble to open to many factions, and that without changing the version of areas)

**Comment by Kalio-42** [comment]
Is it good @SmallJoker ?
Or you don't have time to review ?

**Comment by SmallJoker** [comment]
LGTM. Will merge in a few days unless there are objections.

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**